internals/overlord/servstate/state-diagram.dot

```diff
@@ -6,6 +6,7 @@ digraph service_state_machine {
     node [penwidth=3 shape=box fontsize=24] initial
     node [penwidth=1]
     initial -> starting [label="start"]
+    starting -> backoff [label="exited before okay wait elapses\n(action \"restart\")"]
```

Comment:
I've updated this locally, because I think there were a couple we were missing (and I simplified a couple):

```
digraph service_state_machine {
    node [penwidth=3 shape=box fontsize=24] initial
    node [penwidth=1]

    {backoff, stopped, exited} -> starting [label="start"]

    exited -> stopped [label="stop"]

    initial -> starting [label="start"]

    starting -> running [label="okay wait\nelapsed"]

    starting -> stopped [label="exited"]
    starting -> exited [label="exited"]
    starting -> backoff [label="exited"]
    running -> stopped [label="exited"]
    running -> exited [label="exited"]
    running -> backoff [label="exited"]
    {terminating, killing} -> stopped [label="exited\n(not restarting)"]
    {terminating, killing} -> backoff [label="exited\n(restarting)"]

    starting -> terminating [label="stop (before\nokay wait elapses)"]
    running -> terminating [label="stop"]
    backoff -> stopped [label="stop"]

    backoff -> running [label="backoff time\nelapsed"]

    terminating -> killing [label="terminate time\nelapsed"]

    killing -> stopped [label="kill time\nelapsed"]

    running -> terminating [label="check failed\n(action \"restart\")"]
    exited -> backoff [label="check failed\n(action \"restart\")"]
}
```

However, looking at the output, it's kinda just spaghetti. I'm wondering if a nice hand-drawn ASCII art diagram would be easier to understand, and easier to maintain? It could show the "happy path" (initial -> starting -> running) in a line, with the others off to the side. What do you think -- is the SVG getting out of hand?

Reply:
> I've updated this locally, because I think there were a couple we were missing (and I simplified a couple):
> 
> ```
> digraph service_state_machine {
>     node [penwidth=3 shape=box fontsize=24] initial
>     node [penwidth=1]
> 
>     {backoff, stopped, exited} -> starting [label="start"]
> 
>     exited -> stopped [label="stop"]
> 
>     initial -> starting [label="start"]
> 
>     starting -> running [label="okay wait\nelapsed"]
> 
>     starting -> stopped [label="exited"]
>     starting -> exited [label="exited"]
>     starting -> backoff [label="exited"]
>     running -> stopped [label="exited"]
>     running -> exited [label="exited"]
>     running -> backoff [label="exited"]
>     {terminating, killing} -> stopped [label="exited\n(not restarting)"]
>     {terminating, killing} -> backoff [label="exited\n(restarting)"]
> 
>     starting -> terminating [label="stop (before\nokay wait elapses)"]
>     running -> terminating [label="stop"]
>     backoff -> stopped [label="stop"]
> 
>     backoff -> running [label="backoff time\nelapsed"]
> 
>     terminating -> killing [label="terminate time\nelapsed"]
> 
>     killing -> stopped [label="kill time\nelapsed"]
> 
>     running -> terminating [label="check failed\n(action \"restart\")"]
>     exited -> backoff [label="check failed\n(action \"restart\")"]
> }
> ```
> 
> However, looking at the output, it's kinda just spaghetti. I'm wondering if a nice hand-drawn ASCII art diagram would be easier to understand, and easier to maintain? It could show the "happy path" (initial -> starting -> running) in a line, with the others off to the side. What do you think -- is the SVG getting out of hand?

I have done some investigation on this matter:

## 1 Using [Graphviz](https://graphviz.org/)'s Subgraph and Rank Features

I refactored the current diagram to the following, with ranks and subgraphs:

```
digraph service_state_machine {
    nodesep=1;
    splines=true; // Use splines to create smoother edges and avoid sharp corners.
    node [penwidth=1 shape=box fontsize=14]; // Smaller font size.
    initial [penwidth=3];

    {
        rank=same;  // Forces these nodes onto the same rank
        initial;
        starting;
        running;
    }

    subgraph cluster_intermediate {
        label = "intermediate status";
        style=filled;
        color=lightgrey;
        backoff; 
    }

    subgraph cluster_terminal {
        label="Terminal States";
        style=filled;
        color=lightgrey;
        {
            rank=sink; // Places exited on the rightmost rank.
            exited;
        }
    }

    subgraph cluster_inactive {
        label = "inactive status";
        style=filled;
        color=lightgrey;
        terminating; killing;
    }

    initial -> starting [label="start"];

    starting -> backoff [label="exited before okay wait elapses\n(action \"restart\")"]
    starting -> running [label="okay wait\nelapsed"]
    starting -> terminating [label="stop (before\nokay wait elapses)"]
    starting -> exited [label="exited"]

    running -> terminating [label="stop / check failed (action \"restart\")"]
    running -> exited [label="exited\n(action \"ignore\"/\"shutdown\")"]
    running -> backoff [label="exited\n(action \"restart\")"]

    terminating -> killing [label="terminate time\nelapsed"]
    terminating -> stopped [label="exited\n(not restarting)"]
    terminating -> backoff [label="exited\n(restarting)"]

    exited -> stopped [label="stop"]
    exited -> backoff [label="check failed\n(action \"restart\")"]
    exited -> starting [label="start"]

    backoff -> starting [label="start"]
    backoff -> stopped [label="stop"]
    backoff -> running [label="backoff time\nelapsed"]

    killing -> stopped [label="kill time elapsed / exited (not restarting)"]
    killing -> backoff [label="exited\n(restarting)"]

    stopped -> starting [label="start"]
}
```

Result:

![test](https://github.com/user-attachments/assets/028614fc-38ec-494c-b41a-bd9bdbc03453)

If anything, it was worse than before. It didn't help if I changed the direction of the diagram, either, so I think we might have to give it up.

## 2 [PlantUML](https://github.com/plantuml/plantuml)

I did some research and it seemed PlantUML is also a popular choice for drawing state transfer diagrams, so I gave it a try. It's open-source and has got 10k+ stars on GitHub, so I think it should be a tool that can last for quite a long time.

You can run it locally, but the easiest way is to paste the code and render it on the [online server](https://www.plantuml.com/plantuml/uml). 

After some testing, I ended up with a piece of code like this:

```
@startuml
left to right direction

[*] -> Initial

state "active" as active {
    Initial --> Starting : start
    Starting --> Running : okay wait\nelapsed
}

state "intermediate" as intermediate {
    state Backoff
}

state "inactive" as inactive {
    Terminating --> Killing : terminate time\nelapsed
    Terminating --> Stopped : exited\n(not restarting)
    Killing --> Stopped : exited\n(not restarting)
}

state "terminal" as terminal {
    state Exited
}

Starting ---> Backoff : exited before okay wait elapses\n(action "restart")
Starting --> Terminating : stop (before\nokay wait elapses)
Starting ---> Exited : exited

Running --> Terminating : stop/check failed\n(action "restart")
Running ---> Exited : exited\n(action "ignore"/"shutdown")
Running ---> Backoff :exited\n(action "restart")

Killing ---> Backoff : exited\n(restarting)
Killing --> Stopped : kill time\nelapsed

Terminating ---> Backoff : exited\n(restarting)

Stopped --> Starting : start

Backoff --> Starting : start
Backoff --> Stopped : stop
Backoff --> Running : backoff time\nelapsed

Exited --> Stopped : stop
Exited --> Starting : start
Exited ---> Backoff : check failed\n(action "restart")

@enduml
```

And I have to say the result is a whole lot better:

![state](https://github.com/user-attachments/assets/91cb1221-4dfd-432f-ba67-c6ad094821ec)

It can adjust the length of the lines, which will impact the rendered result, making it slightly better than Graphvis.

If this is acceptable, we can go this way. If this is still less than ideal, the last resort is manual:

## 3 [Excalidraw](https://excalidraw.com/)

Fully manual, with a hand-look feeling. The source file can be downloaded and re-opened. I haven't finished all the drawings yet, just want to see if it could be a choice. An example:

![image](https://github.com/user-attachments/assets/64a195b4-984b-4507-b0db-b52c59c4b561)

Reply:
Per voice discussion, we decided to just get rid of the diagram for now, and if we add this back later, we'd do a nice hand-drawn one for developer documentation (with nice arrows).